### PR TITLE
[flang] Improve debug info for functions.

### PR DESCRIPTION
**Pull request by abidh**
This PR improves the debug information for functions in the following ways:

1. Get line number information from FuncOp and remove hard-coded line numbers.
2. Use proper type for function signature. I have a added a type converter. Currently, it is very limited but will be enhanced with time.
3. Use de-constructed function name.

### CODE REVIEW

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

Would it be possible to reconstruct the type for Complex, Derived and various array types without passing additional info from the frontend?

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 1]
```diff
@@ -106,14 +107,27 @@ void AddDebugInfoPass::runOnOperation() {
...
-        /*encoding=*/1);
-    // FIXME: Provide proper type for subroutine
+    auto result = fir::NameUniquer::deconstruct(funcOp.getName());
+    mlir::StringAttr funcName =
+        mlir::StringAttr::get(context, result.second.name);
+
+    llvm::SmallVector<mlir::LLVM::DITypeAttr> types;
+    fir::DebugTypeGenerator typeGen(module);
+    for (auto resTy : funcOp.getResultTypes()) {
+      auto tyAttr = typeGen.convertType(fir::unwrapRefType(resTy), fileAttr,
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:119]: Why do we need to remove the reference type of returned values?

Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:119]: You are right that it is not required. I will remove it. On this topic, can you kindly explain why the arguments are wrapped in the reference?

**Comment by abidh** [comment]
> LGTM.
> 
> Would it be possible to reconstruct the type for Complex, Derived and various array types without passing additional info from the frontend?

I think complex and array type should be ok. Some work may need to be done for calculation of offsets for derived types but @jeanPerier commented on the design document that it should be do-able using FIR RecordType type directly + mlir datalayout.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 1 above)
Comment (tblah) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:119]: Without the VALUE keyword, fortran defaults to passing all arguments by reference

**Review by tblah — APPROVED** [review]
thanks

**Comment by DavidSpickett** [comment]
This has caused a test suite failure. This particular bot is red already but this specific failure is new:
https://lab.llvm.org/buildbot/#/builders/184/builds/11961
```
> flang-new: ../llvm/mlir/lib/IR/Types.cpp:126: unsigned int mlir::Type::getIntOrFloatBitWidth() const: Assertion `isIntOrFloat() && "only integers and floats have a bitwidth"' failed.
```
The test file is https://github.com/llvm/llvm-test-suite/blob/main/Fortran/gfortran/regression/lto/20091016-1_0.f90.

**Comment by abidh** [comment]
Thanks @DavidSpickett for reverting it. I will investigate the regression.

**Pull request merged**